File: src/server/src/authoring/vocabulary-linter.ts
/**
 * The registry-grounded vocabulary linter.
 *
 * This is the key divergence from the .NET {@link ../../backend/IAW.Vdf.Authoring/Linting/VocabularyLinter.cs}:
 * the .NET linter grounds subject validation on a STATIC `VocabularyCatalog`,
 * whereas this one grounds on the LIVE entity registry projection (N1/N3). The
 * legal subjects are exactly the Active `entity.field` paths the registry knows,
 * and — because the registry is typed — we can additionally catch operator/type
 * mismatches the static catalog could never see (LINT020).
 *
 * Sanity checks (missing outcome params, assert-with-Continue) mirror the .NET
 * linter so the corpus lints identically across both stacks.
 */

import { FieldDataType } from '@prisma/client';

import { ReferenceDataProvider } from '../vdf/reference-data';
import { deserializeRule } from '../vdf/serializer';
import {
  Condition,
  GroupCondition,
  LeafCondition,
  Outcome,
  RecoveryStrategy,
  RuleDefinition,
} from '../vdf/types';

/** Severity of a single lint finding. */
export type FindingSeverity = 'Error' | 'Warning';

/** A single diagnostic produced by the linter. */
export interface LintFinding {
  severity: FindingSeverity;
  /** Machine-readable code (e.g. `LINT001`). */
  code: string;
  /** Human-readable description. */
  message: string;
  /** The logical path within the rule where the issue was found. */
  path: string;
}

/** A complete lint report for one rule. */
export interface LintReport {
  /** True when there are no `Error`-severity findings. */
  isValid: boolean;
  /** Every finding (errors and warnings). */
  findings: LintFinding[];
}

/**
 * A grounded subject the linter validates against: the legal canonical path, its
 * registry data type, and any declared closed value set. This is the projected
 * shape from {@link ../rules/vocabulary-projection.service.GroundedSubject}, kept
 * as a structural type so the pure linter carries no NestJS/Prisma coupling.
 */
export interface GroundingSubject {
  path: string;
  dataType: FieldDataType;
  allowedValues: string[];
}

/** Operators that imply a numeric/ordered comparison (illegal on free String fields). */
const NUMERIC_OPERATORS = new Set<LeafCondition['operator']>([
  'GreaterThan',
  'LessThan',
  'GreaterOrEqual',
  'LessOrEqual',
  'WithinRange',
]);

/**
 * Validates a {@link RuleDefinition} against the registry-projected grounding
 * vocabulary and a reference-data provider. Construct one per grounding snapshot;
 * it is otherwise stateless and deterministic.
 */
export class VocabularyLinter {
  private readonly byPath: Map<string, GroundingSubject>;
  private readonly knownPaths: Set<string>;

  /**
   * @param subjects The grounding vocabulary (registry-projected legal subjects).
   * @param references Reference-data provider used to resolve `reference` keys.
   */
  constructor(
    subjects: readonly GroundingSubject[],
    private readonly references: ReferenceDataProvider,
  ) {
    // The registry canonicalises entity keys to lower-case (its documented
    // case-insensitive identity), so the projected paths carry a lower-case
    // entity segment while field segments keep their authored case. We index by
    // that same normalisation so a corpus subject like `medicalReview.decision`
    // matches the projected `medicalreview.decision` without masking field typos.
    this.byPath = new Map(subjects.map((s) => [normalizeEntity(s.path), s]));
    this.knownPaths = new Set(this.byPath.keys());
  }

  /** Lints an already-deserialized rule. */
  lint(rule: RuleDefinition): LintReport {
    const findings: LintFinding[] = [];

    if (rule.appliesWhen) {
      this.lintCondition(rule.appliesWhen, 'appliesWhen', findings);
    }
    if (rule.assert) {
      this.lintCondition(rule.assert, 'assert', findings);
    }

    this.lintOutcome(rule.onSuccess, 'onSuccess', findings);
    this.lintOutcome(rule.onFailure, 'onFailure', findings);

    if (rule.recover) {
      this.lintRecovery(rule.recover, findings);
    }

    // LINT101 (Warning): asserts a condition but OnFailure is a no-op Continue.
    if (rule.assert && rule.onFailure.type === 'Continue') {
      findings.push({
        severity: 'Warning',
        code: 'LINT101',
        message:
          'Rule asserts a condition but OnFailure is Continue — did you mean to produce a hold or alert?',
        path: 'onFailure',
      });
    }

    const isValid = !findings.some((f) => f.severity === 'Error');
    return { isValid, findings };
  }

  /** Deserializes the JSON then lints; deserialization failures become a LINT000 error. */
  lintJson(ruleJson: string): LintReport {
    let rule: RuleDefinition;
    try {
      rule = deserializeRule(ruleJson);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      return {
        isValid: false,
        findings: [
          {
            severity: 'Error',
            code: 'LINT000',
            message: `Failed to deserialize rule JSON: ${message}`,
            path: '',
          },
        ],
      };
    }
    return this.lint(rule);
  }

  // ── Conditions ────────────────────────────────────────────────────────────

  private lintCondition(
    condition: Condition,
    path: string,
    findings: LintFinding[],
  ): void {
    if (condition.type === 'leaf') {
      this.lintLeaf(condition, path, findings);
      return;
    }
    const group: GroupCondition = condition;
    group.conditions.forEach((child, i) =>
      this.lintCondition(child, `${path}.conditions[${i}]`, findings),
    );
  }

  private lintLeaf(
    leaf: LeafCondition,
    path: string,
    findings: LintFinding[],
  ): void {
    const subject = this.resolveSubject(leaf.subject);

    // LINT001 (Error): unknown subject — not a known entity.field in the registry.
    if (subject === undefined) {
      findings.push({
        severity: 'Error',
        code: 'LINT001',
        message: `Unknown subject '${leaf.subject}' — not a known entity.field in the registry`,
        path: `${path}.subject`,
      });
    }

    // LINT003 (Error): reference key the engine cannot resolve.
    if (leaf.reference !== undefined) {
      if (!this.references.tryResolve(leaf.reference).found) {
        findings.push({
          severity: 'Error',
          code: 'LINT003',
          message: `Unknown reference '${leaf.reference}'`,
          path: `${path}.reference`,
        });
      }
    }

    // LINT020 (Warning): type-aware operator/value mismatch — leverages the
    // registry's declared field type + allowedValues (the typed-registry payoff).
    if (subject !== undefined) {
      this.lintTypeMismatch(leaf, subject, path, findings);
    }
  }

  /**
   * LINT020: registry-type-aware operator/value sanity.
   *  - A numeric/range operator (GreaterThan/LessThan/GreaterOrEqual/LessOrEqual/
   *    WithinRange) applied to a String field is almost certainly a mistake.
   *  - An InSet/Equals against an `allowedValues` field with an inline value
   *    outside the declared set is a mistake.
   */
  private lintTypeMismatch(
    leaf: LeafCondition,
    subject: GroundingSubject,
    path: string,
    findings: LintFinding[],
  ): void {
    if (
      NUMERIC_OPERATORS.has(leaf.operator) &&
      subject.dataType === FieldDataType.String
    ) {
      findings.push({
        severity: 'Warning',
        code: 'LINT020',
        message: `Operator '${leaf.operator}' is numeric/range but subject '${subject.path}' is a String field`,
        path: `${path}.operator`,
      });
      return;
    }

    if (
      (leaf.operator === 'InSet' ||
        leaf.operator === 'NotInSet' ||
        leaf.operator === 'Equals' ||
        leaf.operator === 'NotEquals') &&
      subject.allowedValues.length > 0 &&
      typeof leaf.value === 'string' &&
      !subject.allowedValues.includes(leaf.value)
    ) {
      findings.push({
        severity: 'Warning',
        code: 'LINT020',
        message: `Value '${leaf.value}' is not among the allowed values for '${subject.path}' [${subject.allowedValues.join(', ')}]`,
        path: `${path}.value`,
      });
    }
  }

  // ── Outcomes ──────────────────────────────────────────────────────────────

  private lintOutcome(
    outcome: Outcome,
    path: string,
    findings: LintFinding[],
  ): void {
    // LINT002 (Error): unknown outcome type. The serializer rejects unknown types
    // up-front, but a directly-constructed RuleDefinition could still carry one.
    if (!KNOWN_OUTCOME_TYPES.has(outcome.type)) {
      findings.push({
        severity: 'Error',
        code: 'LINT002',
        message: `Unknown outcome type '${outcome.type}'`,
        path: `${path}.type`,
      });
    }

    requireParam(
      outcome,
      'CreatePlaceholder',
      'SpecimenType',
      'LINT005',
      path,
      findings,
    );
    requireParam(
      outcome,
      'RouteToReview',
      'Destination',
      'LINT006',
      path,
      findings,
    );
    requireParam(outcome, 'PreventAction', 'Action', 'LINT007', path, findings);

    // LINT008 (Error): derivation outcome without a Target.
    if (
      outcome.type === 'SetValue' ||
      outcome.type === 'ApplyDefault' ||
      outcome.type === 'CalculateValue'
    ) {
      if (isBlank(outcome.parameters['Target'])) {
        findings.push({
          severity: 'Error',
          code: 'LINT008',
          message: 'Derivation outcome missing Target parameter',
          path: `${path}.parameters.Target`,
        });
      }
    }

    // LINT102 (Warning): AllowAction without an Action.
    if (
      outcome.type === 'AllowAction' &&
      isBlank(outcome.parameters['Action'])
    ) {
      findings.push({
        severity: 'Warning',
        code: 'LINT102',
        message: 'AllowAction outcome missing Action parameter',
        path: `${path}.parameters.Action`,
      });
    }
  }

  private lintRecovery(
    recover: RecoveryStrategy,
    findings: LintFinding[],
  ): void {
    // LINT004 (Error): recovery references an unresolvable reference key.
    const refVal = recover.parameters['Reference'];
    if (typeof refVal === 'string' && refVal.trim() !== '') {
      if (!this.references.tryResolve(refVal).found) {
        findings.push({
          severity: 'Error',
          code: 'LINT004',
          message: `Recovery references unknown key '${refVal}'`,
          path: 'recover.parameters.Reference',
        });
      }
    }
  }

  // ── Subject resolution ──────────────────────────────────────────────────────

  /**
   * Resolves a subject path to its grounding entry, normalising array fan-out:
   * `order.specimens[].type` falls back to the base collection field
   * `order.specimens[]` (matching the .NET linter's `IsKnownSubjectPath`).
   */
  private resolveSubject(subject: string): GroundingSubject | undefined {
    const normalized = normalizeEntity(subject);
    const direct = this.byPath.get(normalized);
    if (direct !== undefined) {
      return direct;
    }
    const bracket = normalized.indexOf('[]');
    if (bracket >= 0) {
      const base = normalized.slice(0, bracket + 2);
      if (this.knownPaths.has(base)) {
        return this.byPath.get(base);
      }
    }
    return undefined;
  }
}

/**
 * Normalises a subject path's ENTITY (first dot-segment) to lower-case, leaving
 * field segments untouched. Mirrors the registry's canonical key form so subject
 * matching is case-insensitive on the entity (e.g. `medicalReview.decision` →
 * `medicalreview.decision`) without masking case-sensitive field typos.
 */
function normalizeEntity(path: string): string {
  const dot = path.indexOf('.');
  if (dot < 0) {
    return path.toLowerCase();
  }
  return path.slice(0, dot).toLowerCase() + path.slice(dot);
}

/** The closed set of legal outcome types (mirrors {@link OutcomeType}). */
const KNOWN_OUTCOME_TYPES = new Set<string>([
  'Continue',
  'Suppressed',
  'CompleteHold',
  'PartialHold',
  'Warning',
  'ComplianceAlert',
  'RouteToReview',
  'RouteToQueue',
  'Escalate',
  'SetValue',
  'ApplyDefault',
  'CalculateValue',
  'CreatePlaceholder',
  'CreateIncident',
  'CreateTask',
  'PreventAction',
  'AllowAction',
]);

/** True when a parameter value is null/undefined or blank-string. */
function isBlank(value: unknown): boolean {
  return (
    value === undefined ||
    value === null ||
    (typeof value === 'string' && value.trim() === '')
  );
}

/** Adds an Error finding when `outcome` is of `type` but lacks a non-blank `param`. */
function requireParam(
  outcome: Outcome,
  type: Outcome['type'],
  param: string,
  code: string,
  path: string,
  findings: LintFinding[],
): void {
  if (outcome.type === type && isBlank(outcome.parameters[param])) {
    findings.push({
      severity: 'Error',
      code,
      message: `${type} outcome missing ${param} parameter`,
      path: `${path}.parameters.${param}`,
    });
  }
}

File: src/backend/IAW.Vdf.Authoring/Linting/VocabularyLinter.cs
using IAW.Vdf.Abstractions.Conditions;
using IAW.Vdf.Abstractions.Outcomes;
using IAW.Vdf.Abstractions.ReferenceData;
using IAW.Vdf.Abstractions.Rules;
using IAW.Vdf.Abstractions.Vocabulary;
using IAW.Vdf.Core.Serialization;

namespace IAW.Vdf.Authoring.Linting;

/// <summary>Severity of a lint finding.</summary>
public enum FindingSeverity
{
    /// <summary>A potential issue that may indicate a mistake.</summary>
    Warning,

    /// <summary>A definite problem that will likely cause incorrect behaviour.</summary>
    Error
}

/// <summary>A single diagnostic finding from the linter.</summary>
/// <param name="Severity">Severity level of the finding.</param>
/// <param name="Code">Machine-readable code identifying the finding type.</param>
/// <param name="Message">Human-readable description of the issue.</param>
/// <param name="Path">The logical path within the rule where the issue was found.</param>
public sealed record LintFinding(FindingSeverity Severity, string Code, string Message, string Path);

/// <summary>A complete lint report for one rule.</summary>
/// <param name="IsValid">
/// <see langword="true"/> when there are no <see cref="FindingSeverity.Error"/> findings.
/// </param>
/// <param name="Findings">All findings (both errors and warnings).</param>
public sealed record LintReport(bool IsValid, IReadOnlyList<LintFinding> Findings);

/// <summary>
/// Validates a <see cref="RuleDefinition"/> against the <see cref="VocabularyCatalog"/> and an
/// <see cref="IReferenceDataProvider"/>, producing actionable <see cref="LintFinding"/> diagnostics.
/// </summary>
public sealed class VocabularyLinter
{
    private readonly VocabularyCatalog _catalog;
    private readonly IReferenceDataProvider _references;

    /// <summary>
    /// Initializes a new <see cref="VocabularyLinter"/>.
    /// </summary>
    /// <param name="catalog">The vocabulary catalog to validate subjects and outcome types against.</param>
    /// <param name="references">The reference data provider to resolve reference keys against.</param>
    public VocabularyLinter(VocabularyCatalog catalog, IReferenceDataProvider references)
    {
        _catalog = catalog;
        _references = references;
    }

    /// <summary>
    /// Lints an already-deserialized <see cref="RuleDefinition"/>.
    /// </summary>
    /// <param name="rule">The rule to lint.</param>
    /// <returns>A <see cref="LintReport"/> containing all findings.</returns>
    public LintReport Lint(RuleDefinition rule)
    {
        var findings = new List<LintFinding>();

        // Walk all conditions for subject and reference validation.
        if (rule.AppliesWhen is not null)
            LintCondition(rule.AppliesWhen, "appliesWhen", findings);

        if (rule.Assert is not null)
            LintCondition(rule.Assert, "assert", findings);

        // Lint outcomes.
        LintOutcome(rule.OnSuccess, "onSuccess", findings);
        LintOutcome(rule.OnFailure, "onFailure", findings);

        // Lint recovery.
        if (rule.Recover is not null)
            LintRecovery(rule.Recover, findings);

        // LINT101: Assert with trivial OnFailure.
        if (rule.Assert is not null && rule.OnFailure.Type == OutcomeType.Continue)
        {
            findings.Add(new LintFinding(
                FindingSeverity.Warning,
                "LINT101",
                "Rule asserts a condition but OnFailure is Continue — did you mean to produce a hold or alert?",
                "onFailure"));
        }

        // LINT110: Authored scope vs. referenced subjects (non-blocking).
        // Only runs when the author declared a Scope; scopeless rules (Scope == null) are skipped,
        // so the shipped corpus (which has no scope) is unaffected.
        if (rule.Scope is not null)
            LintScope(rule, findings);

        var isValid = !findings.Any(f => f.Severity == FindingSeverity.Error);
        return new LintReport(isValid, findings);
    }

    /// <summary>
    /// Deserializes the JSON then lints; catches JSON deserialization errors as findings.
    /// </summary>
    /// <param name="ruleJson">The JSON text to deserialize and lint.</param>
    /// <returns>A <see cref="LintReport"/> containing all findings.</returns>
    public LintReport LintJson(string ruleJson)
    {
        RuleDefinition rule;
        try
        {
            rule = RuleSerializer.Deserialize(ruleJson);
        }
        catch (Exception ex)
        {
            var finding = new LintFinding(
                FindingSeverity.Error,
                "LINT000",
                $"Failed to deserialize rule JSON: {ex.Message}",
                "");
            return new LintReport(false, new[] { finding });
        }

        return Lint(rule);
    }

    // ── Private helpers ──────────────────────────────────────────────────────────────────────────

    private void LintCondition(ICondition condition, string path, List<LintFinding> findings)
    {
        switch (condition)
        {
            case LeafCondition leaf:
                LintLeafCondition(leaf, path, findings);
                break;

            case GroupCondition group:
                for (var i = 0; i < group.Conditions.Count; i++)
                    LintCondition(group.Conditions[i], $"{path}.conditions[{i}]", findings);
                break;
        }
    }

    private void LintLeafCondition(LeafCondition leaf, string path, List<LintFinding> findings)
    {
        // LINT001: Unknown subject path.
        if (!IsKnownSubjectPath(leaf.Subject))
        {
            findings.Add(new LintFinding(
                FindingSeverity.Error,
                "LINT001",
                $"Unknown subject '{leaf.Subject}'",
                $"conditions.{leaf.Subject}"));
        }

        // LINT003: Unknown reference key.
        if (leaf.Reference is not null)
        {
            if (!_catalog.IsKnownReference(leaf.Reference) &&
                !_references.TryResolve(leaf.Reference, out _))
            {
                findings.Add(new LintFinding(
                    FindingSeverity.Error,
                    "LINT003",
                    $"Unknown reference '{leaf.Reference}'",
                    $"{path}.reference"));
            }
        }
    }

    private void LintOutcome(Outcome outcome, string path, List<LintFinding> findings)
    {
        // LINT002: Unknown outcome type (all types are in the default catalog, so this fires only for
        // catalogs that don't include all types).
        if (!_catalog.IsKnownOutcome(outcome.Type))
        {
            findings.Add(new LintFinding(
                FindingSeverity.Error,
                "LINT002",
                $"Unknown outcome type '{outcome.Type}'",
                $"{path}.type"));
        }

        // LINT005: CreatePlaceholder without SpecimenType.
        if (outcome.Type == OutcomeType.CreatePlaceholder)
        {
            if (!outcome.Parameters.TryGetValue("SpecimenType", out var st) ||
                st is null || string.IsNullOrWhiteSpace(st?.ToString()))
            {
                findings.Add(new LintFinding(
                    FindingSeverity.Error,
                    "LINT005",
                    "CreatePlaceholder outcome missing SpecimenType parameter",
                    $"{path}.parameters.SpecimenType"));
            }
        }

        // LINT006: RouteToReview without Destination.
        if (outcome.Type == OutcomeType.RouteToReview)
        {
            if (!outcome.Parameters.TryGetValue("Destination", out var dest) ||
                dest is null || string.IsNullOrWhiteSpace(dest?.ToString()))
            {
                findings.Add(new LintFinding(
                    FindingSeverity.Error,
                    "LINT006",
                    "RouteToReview outcome missing Destination parameter",
                    $"{path}.parameters.Destination"));
            }
        }

        // LINT007: PreventAction without Action.
        if (outcome.Type == OutcomeType.PreventAction)
        {
            if (!outcome.Parameters.TryGetValue("Action", out var action) ||
                action is null || string.IsNullOrWhiteSpace(action?.ToString()))
            {
                findings.Add(new LintFinding(
                    FindingSeverity.Error,
                    "LINT007",
                    "PreventAction outcome missing Action parameter",
                    $"{path}.parameters.Action"));
            }
        }

        // LINT008: Derivation outcome without Target.
        if (outcome.Type is OutcomeType.SetValue or OutcomeType.ApplyDefault or OutcomeType.CalculateValue)
        {
            if (!outcome.Parameters.TryGetValue("Target", out var target) ||
                target is null || string.IsNullOrWhiteSpace(target?.ToString()))
            {
                findings.Add(new LintFinding(
                    FindingSeverity.Error,
                    "LINT008",
                    "Derivation outcome missing Target parameter",
                    $"{path}.parameters.Target"));
            }
        }

        // LINT102: AllowAction without Action (warning).
        if (outcome.Type == OutcomeType.AllowAction)
        {
            if (!outcome.Parameters.TryGetValue("Action", out var action) ||
                action is null || string.IsNullOrWhiteSpace(action?.ToString()))
            {
                findings.Add(new LintFinding(
                    FindingSeverity.Warning,
                    "LINT102",
                    "AllowAction outcome missing Action parameter",
                    $"{path}.parameters.Action"));
            }
        }
    }

    private void LintRecovery(RecoveryStrategy recover, List<LintFinding> findings)
    {
        // LINT004: Reference key in recovery parameters that cannot be resolved.
        if (recover.Parameters.TryGetValue("Reference", out var refVal) &&
            refVal is string refKey &&
            !string.IsNullOrWhiteSpace(refKey))
        {
            if (!_catalog.IsKnownReference(refKey) &&
                !_references.TryResolve(refKey, out _))
            {
                findings.Add(new LintFinding(
                    FindingSeverity.Error,
                    "LINT004",
                    $"Recovery references unknown key '{refKey}'",
                    "recover.parameters.Reference"));
            }
        }
    }

    /// <summary>
    /// LINT110 (Warning, non-blocking): cross-checks the author-declared <see cref="RuleScope"/> against
    /// the subjects the rule actually references. A finding is raised when a referenced subject's object
    /// (the first dotted segment, with a trailing <c>[]</c> stripped) is not declared in
    /// <see cref="RuleScope.Objects"/>, or — when <see cref="RuleScope.Properties"/> is non-empty — when a
    /// referenced subject path is not among the declared properties. Never produces an Error, so it cannot
    /// reject a rule; it surfaces drift between the declared scope and the rule body for governance review.
    /// </summary>
    private void LintScope(RuleDefinition rule, List<LintFinding> findings)
    {
        var scope = rule.Scope!;
        var referencedSubjects = new List<string>();
        CollectSubjects(rule.AppliesWhen, referencedSubjects);
        CollectSubjects(rule.Assert, referencedSubjects);

        if (referencedSubjects.Count == 0)
            return;

        var scopedObjects = new HashSet<string>(scope.Objects, StringComparer.Ordinal);
        var scopedProperties = new HashSet<string>(scope.Properties, StringComparer.Ordinal);
        var checkProperties = scopedProperties.Count > 0;

        foreach (var subject in referencedSubjects.Distinct(StringComparer.Ordinal))
        {
            var objectName = ObjectName(subject);

            if (!scopedObjects.Contains(objectName))
            {
                findings.Add(new LintFinding(
                    FindingSeverity.Warning,
                    "LINT110",
                    $"Rule references object '{objectName}' (subject '{subject}') which is outside its declared scope objects [{string.Join(", ", scope.Objects)}].",
                    "scope.objects"));
                continue;
            }

            if (checkProperties && !scopedProperties.Contains(subject))
            {
                findings.Add(new LintFinding(
                    FindingSeverity.Warning,
                    "LINT110",
                    $"Rule references subject '{subject}' which is outside its declared scope properties.",
                    "scope.properties"));
            }
        }
    }

    /// <summary>Collects every leaf subject path referenced under a condition (recursively).</summary>
    private static void CollectSubjects(ICondition? condition, List<string> into)
    {
        switch (condition)
        {
            case null:
                return;
            case LeafCondition leaf:
                if (!string.IsNullOrWhiteSpace(leaf.Subject))
                    into.Add(leaf.Subject);
                break;
            case GroupCondition group:
                foreach (var child in group.Conditions)
                    CollectSubjects(child, into);
                break;
        }
    }

    /// <summary>
    /// The object name for a subject path: the first dotted segment with any trailing <c>[]</c>
    /// stripped (e.g. <c>"order.product"</c> → <c>"order"</c>, <c>"order.tests[]"</c> → <c>"order"</c>,
    /// <c>"specimens[].type"</c> → <c>"specimens"</c>).
    /// </summary>
    private static string ObjectName(string subject)
    {
        var dot = subject.IndexOf('.');
        var head = dot < 0 ? subject : subject[..dot];

        var bracket = head.IndexOf("[]", StringComparison.Ordinal);
        return bracket < 0 ? head : head[..bracket];
    }

    /// <summary>
    /// Checks whether a subject path is known in the catalog.
    /// Handles array fan-out patterns: "order.specimens[].type" → checks "order.specimens[]" base.
    /// </summary>
    private bool IsKnownSubjectPath(string subject)
    {
        if (_catalog.IsKnownSubject(subject))
            return true;

        // If the path contains "[]", try the base collection path (everything up to and including "[]").
        var bracketIdx = subject.IndexOf("[]", StringComparison.Ordinal);
        if (bracketIdx >= 0)
        {
            var basePath = subject[..(bracketIdx + 2)]; // includes the "[]"
            if (_catalog.IsKnownSubject(basePath))
                return true;
        }

        return false;
    }
}
